Add EIP: Asset-bound Non-Fungible Tokens (#6956)
ethereum/EIPs: EIPS/eip-6956.md

---
eip: 6956
title: Asset-bound Non-Fungible Tokens
description: Asset-bound NFTs anchor a token 1-1 to an asset and transfers are authorized through oracle-attestation of control over the asset
author: Thomas Bergmueller (@tbergmueller) <thomas.bergmueller@authenticvision.com>, Lukas Meyer (@ibex-technology) <lukas@ibex.host>
discussions-to: https://ethereum-magicians.org/t/erc-6956-asset-bound-non-fungible-tokens/14056
status: Draft
type: Standards Track
category: ERC
created: 2023-04-29
requires: 165, 721
---

## Abstract

This standard allows to integrate physical and digital ASSETS without signing capabilities into dApps/web3 by extending [ERC-721](eip-721.md).

> An `ASSET`, e.g. a physical object, is equipped with an `ANCHOR`. The `ANCHOR-TECHNOLOGY` must be chosen s.t. an ANCHOR allows to uniquely identify the ASSET. The ANCHOR-TECHNOLOGY must further allow to establish a `PROOF-OF-CONTROL` over the ASSET through an `ORACLE`. For physical ASSETS, PROOF-OF-CONTROL corresponds for example to proof of physical presence.

Reviewer: We generally use backticks to signify code, would you mind switching these to use bold or italics instead? Only because inline code gets some special treatment from the spell checker, etc.

Author: I went with capitals only and removed the backticks. Resolved with https://github.com/ethereum/EIPs/commit/9d188578511cb53c00ebe0910043ab50fb40a248

> An `ASSET`, e.g. a physical object, is equipped with an `ANCHOR`. The `ANCHOR-TECHNOLOGY` must be chosen s.t. an ANCHOR allows to uniquely identify the ASSET. The ANCHOR-TECHNOLOGY must further allow to establish a `PROOF-OF-CONTROL` over the ASSET through an `ORACLE`. For physical ASSETS, PROOF-OF-CONTROL corresponds for example to proof of physical presence.

Reviewer: Please expand "s.t." and for all other abbreviations, make sure to expand them the first time they're used.

Author: done in https://github.com/ethereum/EIPs/commit/9d188578511cb53c00ebe0910043ab50fb40a248

This standard to proposes to use `ATTESTATIONS` as authorization for the following [ERC-721](eip-721.md) mechanisms: `transfer`, `burn` and `approve`. The proposed `transferAnchor(attestation)`, `burnAnchor(attestation)` and `approveAnchor(attestation)` are permissionless, i.e. neither the sender/owner (`from`) nor the receiver (`to`) need to sign. Authorization is solely provided through the ORACLE's ATTESTATION.

We also outline for optional use

- a `FLOATING`-concept (temporarily or permanently enabling "traditional" [ERC-721](eip-721.md) transfers without ATTESTATION)
- `ATTESTATION-LIMITS`, which are recommended to implement for security reasons, when gas is paid through a central account (see Figure 1)

Figure 1 below shows a the data flow of an asset-bound NFT transfer through a simplified example system employing the proposed standard. The system is utilizing a smartphone as user-device to interact with a physical ASSET.

![Figure 1: Sample system](../assets/eip-6956/img/erc6956_concept.svg)

> 

Reviewer: Your abstract is a tad on the long side.

Author: As I mentioned in Office Hours #17 its a bit hard to shorten since the EIP itself is long and slightly complex. Managed to reduce from 250 words to 199 in https://github.com/ethereum/EIPs/commit/9d188578511cb53c00ebe0910043ab50fb40a248 .

Transactions authorized via `ATTESTATION` shall not require signature or approval from neither the `from` (donor, owner, sender) nor `to` (beneficiary, receiver) account, i.e. making transfers permissionless. Ideally, transaction are signed independent from the `ORACLE` as well, allowing different scenarios in terms of gas-fees.

Lastly we want to mention two major side-benefits of using the proposed standard, which drastically lowers hurdles in onboarding web2 users and increase their security;

- New users, e.g `0xaa...aa` (Fig.1), can use gasless wallets, hence participate in Web3/dApps/DeFi and mint+transfer tokens without ever owning crypto currency. Gas-fees may be paid through a third-party account `0x..gasPayer` (Fig.1). The gas is typically covered by the ASSET issuer, who signs `transferAnchor()` transactions

> - Users cannot get scammed. Common attacks (e.g. wallet-drainer scams) are no longer possible or easily reverted, since only the anchored NFT can be stolen, not the ASSET itself. Also mishaps like transferring the NFT to the wrong account, losing access to an account etc can be mitigated by executing another `transferAnchor()` transaction based on proofing control over the `ASSET`, i.e. the physical object.

Reviewer 2: Q: is this putting physical asset thief in capacity to change the NFT property to their benefit?

Author: Yes - it really maps the physical world 1:1, including the not so nice things. So you need to protect the physical asset as you'd do with any other physical thing you own. 

That being said, there is flexibility. As outlined, the ERC-6956 can easily be extended with a lockable-mechanism, which locks attestation transfers. This is important for e.g. DeFi, where the asset is collateral, but could also be used for theft protection. Since the proposed standard is completely compatible with ERC-721, any "standard lock mechanism for NFTs" can be used. If you just want to lock attestations (e.g. when using IERC6956Floatable), you can implement logic in the `_beforeAttestationUse` hook as suggested in the Reference Implementation.

Author: Added this question to *Security Considerations* in https://github.com/ethereum/EIPs/commit/9d188578511cb53c00ebe0910043ab50fb40a248 - Thank you!

- RECOMMENDED to implement any or multiple of the following interfaces: transferable(tokenId), isSoulbound(tokenId), isNonTransferable (), `isNonTransferable(tokenId)` according to [ERC-6454](eip-6454.md) ...

- MAY implement the `IERC6956AttestationLimited` interface subject to [the proposed AttestationLimited extension](#ierc6956attestationlimited)
-- This is a MUST when transaction-costs are provided through a central account, e.g. through the ORACLE (or associated authorities) itself to avoid fund-draining.

- MAY implement the `IERC6956Floatable` interface subject to [the proposed Floatable extension](#ierc6956floatable).

### `IERC6956AttestationLimited`

WIP and needs discussion!!!

Every contract compliant to this standard MAY implement the [proposed AttestationLimited interface](../assets/eip-6956/contracts/IERC6956AttestationLimited.sol) and MUST implement ERC-721 and ERC-165 interfaces (subject to “caveats” below):

> ```Solidity

Reviewer: ````suggestion
```solidity
````

Author: done in https://github.com/ethereum/EIPs/commit/9d188578511cb53c00ebe0910043ab50fb40a248

- New users, e.g `0xaa...aa` (Fig.1), can use gasless wallets, hence participate in Web3/dApps/DeFi and mint+transfer tokens without ever owning crypto currency. Gas-fees may be paid through a third-party account `0x..gasPayer` (Fig.1). The gas is typically covered by the ASSET issuer, who signs `transferAnchor()` transactions
- Users cannot get scammed. Common attacks (for example wallet-drainer scams) are no longer possible or easily reverted, since only the anchored NFT can be stolen, not the ASSET itself. Also mishaps like transferring the NFT to the wrong account, losing access to an account etc can be mitigated by executing another `transferAnchor()` transaction based on proofing control over the ASSET, namely the physical object.

### Related work

We primarily aim to onboard physical or digital ASSETS into dApps, which do not signing-capabilities of their own (contrary to [ERC-5791](eip-5791.md) approach using crypto-chip based solutions). Note that we do not see any restrictions preventing to use ERC-5791 in combination with this standard, as the address of the crypto-chip qualifies as an ANCHOR.

> --- TO BE EXTENDED ---

Reviewer: ```suggestion
<!-- TO BE EXTENDED -->
```

- The ATTESTATION MUST abi-encode the following:
  - `to`, MUST be address, specifying the beneficiary, for example the to-address, approved account etc.
  - ANCHOR, aka the ASSET identifier, MUST have a 1:1 relation to the ASSET
  - `attestationTime`, UTC seconds, time when attestation was signed by ORACLE,
  - `validStartTime` UTC seconds, start time of the ATTESTATION's validity timespan
  - `validEndTime`, UTC seconds, end time of the ATTESTATION's validity timespan
  - `signature`, ETH-signature (65 bytes). Output of an ORACLE signing the `attestationHash = keccak256([to, anchor, attestationTime, validStartTime, validEndTime])`.
- How PROOF-OF-CONTROL is establish in detail through an ANCHOR-TECHNOLOGY is not subject to this standard. Some ORACLE requirements and ANCHOR-TECHNOLOGY requirements when using PHYSICAL ASSETS are outlined in [Security considerations for Physical Assets](#security-considerations-for-physical-assets).

Minimal Typescript sample to generate an ATTESTATION using ethers library:

> ```TypeScript

Reviewer: ````suggestion
```typescript
````

---
eip: 6956
title: Asset-bound Non-Fungible Tokens
description: Asset-bound NFTs anchor a token 1-1 to an asset and operations are authorized through oracle-attestation of control over the asset
author: Thomas Bergmueller (@tbergmueller) <thomas.bergmueller@authenticvision.com>, Lukas Meyer (@ibex-technology) <lukas@ibex.host>
discussions-to: https://ethereum-magicians.org/t/erc-6956-asset-bound-non-fungible-tokens/14056
status: Draft
type: Standards Track
category: ERC
created: 2023-04-29
requires: 165, 721
---

## Abstract

> This standard allows to integrate physical and digital ASSETS without signing capabilities into dApps/web3 by extending [ERC-721](eip-721.md).

Reviewer: ```suggestion
This standard allows integrating physical and digital ASSETS without signing capabilities into dApps/web3 by extending [ERC-721](eip-721.md).
```

### The Problem

ERC-721 outlines that "NFTs can represent ownership over digital or physical assets". ERC-721 excels in this task when used to represent ownership over digital, on-chain assets, that is when the asset is "holding a token of a specific contract" or the asset is an NFT's metadata. Today, people commonly treat an NFT's metadata (images, traits, ...) as asset-class, with their rarity often directly defining the value of an invididual NFT. 

However, we see integrity issues not solveable with ERC-721, primarily when NFTS are used to represent off-chain ASSETS ("ownership over physical products", "digital collectables", "in-game assets", "responsibilities", ...). Over an ASSET's lifecycle, the ASSET's ownership and posession state changes multiple, sometimes thousands, of times. Each of those state changes may result in shifting obligations and privileges for the involved parties. Therefore tokenization of an ASSET *without* enforcably anchoring the ASSET's associated obligation and properties to the token is not complete. Nowadays, off-chain ASSETs are often "anchored" through adding an ASSET-identifier to a NFT's metadata. 

> **NFT-ASSET integrity:** Contrary to common misconception from many NFT-investors, metadata is data that is firstly mutable and secondly off-chain. Therefore the link between an ASSET through an asset-identifier stored in mutable metadata, which is only linked to the NFT through tokenURI, can be considered weak at best.

Reviewer: Although uncommon, it is [possible to embed metadata directly on-chain](https://etherscan.io/address/0xb7d01ca53ffcd0dc11c2a446e49408ded9b15372#code#F1#L268).

```suggestion
**NFT-ASSET integrity:** Contrary to a popular belief among NFT-investors, metadata is data that is, more often than not, mutable and off-chain. Therefore the link between an ASSET through an asset-identifier stored in mutable metadata, which is only linked to the NFT through `tokenURI`, can be considered weak at best.
```

Author: adapted in https://github.com/ethereum/EIPs/pull/6956/commits/1713073d9c204829c446afffa7afd4fbd7754bd6

- Users cannot get scammed. Common attacks (for example wallet-drainer scams) are no longer possible or easily reverted, since only the anchored NFT can be stolen, not the ASSET itself. Also mishaps like transferring the NFT to the wrong account, losing access to an account etc can be mitigated by executing another `transferAnchor()` transaction based on proofing control over the ASSET, namely the physical object.

### Related work

We primarily aim to onboard physical or digital ASSETS into dApps, which do not signing-capabilities of their own (contrary to [ERC-5791](eip-5791.md) approach using crypto-chip based solutions). Note that we do not see any restrictions preventing to use ERC-5791 in combination with this standard, as the address of the crypto-chip qualifies as an ANCHOR.

--- TO BE EXTENDED ---

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

### Definitions (alphabetical)

> - **ANCHOR** uniquely identifies the off-chain ASSET, being it physical or digital.

Reviewer: ```suggestion
- **ANCHOR** uniquely identifies the off-chain ASSET, whether it is physical or digital.
```

Author: Adapted in https://github.com/ethereum/EIPs/pull/6956/commits/1713073d9c204829c446afffa7afd4fbd7754bd6

|---------------|---|---|---|---|---|
| **Managing Posession** |
| Token gating  | ASSET | ANY | incompatible | - | - |
| Digital twin  | ASSET | ANY | incompatible | - | - |
| Scarce digital twin | ASSET | ANY | incompatible | required | - |
| Lendable digital twin         | OWNER_AND_ASSET | ASSET | required | - | - |
| **Managing Ownership** |
| Securing ownership from theft   | OWNER or OWNER_AND_ASSET | ANY | optional | - | required |
| Selling an house with a mortage  | ASSET  or OWNER_AND_ASSET | ANY | optional | optional | required |
| Selling a house with a lease | ASSET or OWNER_AND_ASSET | ANY | optional | optional | required |
| Buying a brand new car with downpayment | ASSET or OWNER_AND_ASSET | ANY | optional | optional | required |
| Buying a barrel of oil by forward transaction | ASSET or OWNER_AND_ASSET | ANY | optional | optional | required |

Legend:

- required ... we don't see a way how to implement the use-case without it
- incompatible ... this MUSTN'T be implemented, as it is a security risk for the use-case
- optional ... this MAY optionally be implemented

## Backwards Compatibility

> ---- NEEDS FURTHER DISCUSSION ----

Reviewer: If you use a comment for these notes, the linting tools will make sure you remove them before moving into review:

```suggestion
<!-- NEEDS FURTHER DISCUSSION -->
```

Author: Done in https://github.com/ethereum/EIPs/pull/6956/commits/1713073d9c204829c446afffa7afd4fbd7754bd6

Thx for the hint with linter!

However, it is RECOMMENDED to extend implementations of the proposed standard with an interface indicating transferability of NFTs for market places. Examples include [ERC-6454](eip-6454.md) and [ERC-5484](eip-5484.md).

Many ERC-721 extensions suggest to add additional throw-conditions to transfer methods. This standard is fully compatible, as

- The often-used ERC-721 `_beforeTokenTransfer()` hook must be called for all transfers including attestation-authorized transfers.
- A `_beforeAnchorUse()` hook is suggested in the reference implementation, which only is called when using attestation as authorization.

## Test Cases

Test cases are available:

- For only implementing [the proposed standard interface](../assets/eip-6956/contracts/IERC6956.sol) can be found [here](../assets/eip-6956/test/ERC6956.ts)
- For implementing [the proposed standard interface](../assets/eip-6956/contracts/IERC6956.sol), [the Floatable extension](../assets/eip-6956/contracts/IERC6956Floatable.sol) and [the AttestationLimited extension](../assets/eip-6956/contracts/IERC6956AttestationLimited.sol) can be found [here](../assets/eip-6956/test/ERC6956Full.ts)

## Reference Implementation

> The reference implementations are [MIT](../assets/eip-6956/LICENSE.md) licensed. They can be freely used with digital assets as defined in the present EIP. No general license is granted to the *commercial* use or application of the reference implementation or derived works with physical assets as defined in the present EIP. Visit metaanchor.io for further details. 

Reviewer: Normally we don't note licensing information in the EIP itself. You're free to include the MIT license in your assets directory though.

---

I am not a lawyer, but if the reference implementation is available under the MIT license, then it _is_ licensed for commercial use as long as you comply with the terms of the MIT license, no?

---

While [we're still working out the specifics](https://ethereum-magicians.org/t/licences-for-assets-reference-implementations-in-eips/14329), we only allow [OSI Approved Licenses](https://opensource.org/licenses/). MIT with this non-commercial requirement is not an approved license.

---

Please remove the link to `metaanchor.io`.

Author: No problem for me to remove it (will do!). 

However and adding to the [forum discussion](https://ethereum-magicians.org/t/licences-for-assets-reference-implementations-in-eips/14329) (will copy it there as well)

To include that limitation was recommendation from our lawyer though. Not to protect us as the authors but mainly to protect users of the reference implementation. Let me quickly outline our intentions;

- For disclosure, we (Authentic Vision, the company I'm working for) offer with our MetaAnchor-Technology (featuring a random hologram and smartphone app, can be seen [in 10sec here](https://youtu.be/bz-sgusrcVA?t=52) ) a setup that is suitable to act as anchor-technology and oracle. We also own patents for that technology.
- Still we want everyone (commercially and non-commercially) to be able to use the reference implementation without limits.
- The rationale of the lawyer was, that many anchor-technologies (such as our MetaAnchor, NFCs, special QR-Codes, ...) will likely be patent-protected. He recommended to add (especially for commercial) a clause to remind people to be mindful when using the reference implementation in combination with physical assets and physical anchor technologies. Just because they're using the reference implementation does not grant them the right to use any arbitrary anchor technology. 
- The Reference Implementation is MIT, because some of our patents may have similarities to parts of the EIP in some embodiments - we want to be very clear that people can use the EIP and reference implementation freely and unlimited, even though parts may be mentioned in our patents.

==> I would argue that if we were to enforce CC0 for reference implementations without amendments, I would have no legal framework to ensure people it is safe to use our reference implementation here. That's why the reference implementation must be MIT or similar imo.

Author: Removed in https://github.com/ethereum/EIPs/pull/6956/commits/1713073d9c204829c446afffa7afd4fbd7754bd6

Author: Neither is possible unfortunately
- I cannot list patents I'm aware of, as I would - depending on which legislation we look at - be liable for the completeness of this list. Also, it would be legal advice, which I cannot give. 
- Patents are typically pending for years, including many of ours. So even if I wanted to, I cannot grant licenses on the grounds of a not-yet granted patent. 

With this whole thing we wanted to reassure people and at the same time raise awareness, that they could be infringing other patents (not ours) especially on the physical side.

But I think maybe we may be seeing things too complicated or hypothetical here. Simply licensing the reference implementation under MIT should do the job and be the most reassuring thing we can do. What do you think?

Author: Following discussion in Discord, should be resolved imho via https://github.com/ethereum/EIPs/commit/830195450c3883a1f9541840d1f1b8ffeb5ae140

- Minimal implementation, only supporting [the proposed standard interface](../assets/eip-6956/contracts/IERC6956.sol) can be found [here](../assets/eip-6956/contracts/ERC6956.sol)

> - Full implementation, support [the proposed standard interface](../assets/eip-6956/contracts/IERC6956.sol), [the Floatable extension](../assets/eip-6956/contracts/IERC6956Floatable.sol) and [the AttestationLimited extension](../assets/eip-6956/contracts/IERC6956AttestationLimited.sol) can be found [here](../assets/eip-6956/contracts/ERC6956Full.sol)

Reviewer: ```suggestion
- Full implementation, with support for [the proposed standard interface](../assets/eip-6956/contracts/IERC6956.sol), [the Floatable extension](../assets/eip-6956/contracts/IERC6956Floatable.sol) and [the AttestationLimited extension](../assets/eip-6956/contracts/IERC6956AttestationLimited.sol) can be found [here](../assets/eip-6956/contracts/ERC6956Full.sol)
```

## Security Considerations

> --- NEEDS FURTHER DISCUSSION ---

Reviewer: ```suggestion
<!-- NEEDS FURTHER DISCUSSION -->
```

#### Physical ASSET

- MUST comprise an ANCHOR, acting as the unique physical object identifier, typically a serial number (plain (NOT RECOMMENDED) or hashed (RECOMMENDED))
- MUST comprise a physical security device, marking or any other feature that enables proofing physical presence for ATTESTATION through the ORACLE
- Is RECOMMENDED to employ ANCHOR-TECHNOLOGIES featuring irreproducible security features.
- In general it is NOT RECOMMENDED to employ ANCHOR-TECHNOLOGIES that can easily be replicated (for example barcodes, "ordinary" NFC chips, .. ). Replication includes physical and digital replication.

### Security Considerations for DIGITAL ASSETS

--- NEEDS DISCUSSION ---

- TODO formal definition, shall extend to digital-only AND abstract assets (for example memberships)
- Maybe use Physical and non-physical assets (to extend to abstract assets)?
- Requirements on Oracles and Proof-of-Control for Digital ASSETs will not be defined in this EIP

Needs discussion.

Rough considerations and TODO

- Valid anchors
  - Outline merkle-tree salt leaves
  - Why using merkle-trees and not simply store all available anchors on-chain (besides memory issues)

> - Maintainance-role over using ownership (Ownable is used by marketplaces to manage the collection)

Author: Done in https://github.com/ethereum/EIPs/pull/6956/commits/1713073d9c204829c446afffa7afd4fbd7754bd6

## Copyright

Copyright and related rights waived via [CC0](../LICENSE.md).

> The reference implementations are [MIT](../assets/eip-6956/LICENSE.md) licensed. They can be freely used with digital assets as defined in the present EIP. No general license is granted to the *commercial* use or application of the reference implementation or derived works with physical assets as defined in the present EIP. Visit metaanchor.io for further details.

Author: Removed in https://github.com/ethereum/EIPs/pull/6956/commits/1713073d9c204829c446afffa7afd4fbd7754bd6
